### PR TITLE
DMP-4194: Implement batching for most automated tasks

### PR DESCRIPTION
**Pull request by Ben-Edwards-cgi**
### Change description ###
At present, batching is only set on a handful of automated tasks. However, in order to ensure the service remains performant, it needs to be ensured that batching is set and used when these tasks run (except Processing Daily List task). 

This ticket will involve:
 * Setting the batch size of automated tasks as per the [LLD|https://tools.hmcts.net/confluence/pages/viewpage.action?pageId=1689786831#TaskautomationLLD-Automatedtasksummary]
 * Updating the cron schedule to align with the [LLD|https://tools.hmcts.net/confluence/pages/viewpage.action?pageId=1689786831#TaskautomationLLD-Automatedtasksummary] (note, if the task is currently set to run 'in hours' and we're changing to now be out of hours, ensure the cron is overridden in the DB on the lower environments to make sure it runs before the environments are shut down). 
 * Ensure all relevant automated tasks actually use the configured batch size. It has been noted as part of performance testing that some automated tasks do not take into account the batch size when set in the DB. See [here|https://tools.hmcts.net/confluence/pages/viewpage.action?pageId=1803676924#DARTSAutomatedTaskChecks-AutomatedTaskPerformanceTests] under 'Batch ignored' column for a list of the tasks where this applies. 

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/darts/task/runner/impl/ApplyRetentionAutomatedTask.java [hunk 1]
```diff
@@ -32,8 +34,13 @@ public AutomatedTaskName getAutomatedTaskName() {
         return APPLY_RETENTION_TASK_NAME;
     }
 
+    @Override
+    public Duration getLockAtMostFor() {
+        return Duration.ofMinutes(90);
```
Comment (chrisbellingham-hmcts) [inline src/main/java/uk/gov/hmcts/darts/task/runner/impl/ApplyRetentionAutomatedTask.java:39]: As a general comment, would there be value in making these lock durations configurable via spring properties? Might be a useful thing to have easily tweakable without requiring a new code release?

File: src/main/java/uk/gov/hmcts/darts/common/repository/CaseRepository.java [hunk 2]
```diff
@@ -23,11 +22,11 @@ Optional<CourtCaseEntity> findByCaseNumberAndCourthouse(String caseNumber,
...
-        SELECT case.caseNumber
-        FROM CourtCaseEntity case
-        WHERE case.closed = false
-        and case.caseNumber in :caseNumbers
-        and case.courthouse.courthouseName = upper(:courthouseName)
+        SELECT c.caseNumber
+        FROM CourtCaseEntity c
+        WHERE c.closed = false
+        and c.caseNumber in :caseNumbers
+        and c.courthouse.courthouseName = upper(:courthouseName)
```
Comment (chrisbellingham-hmcts) [inline src/main/java/uk/gov/hmcts/darts/common/repository/CaseRepository.java:29]: It's a general gripe I have with how most of our JPQL is already written, but out of interest, what's the benefit of changing to single char var names here versus a more verbose name? I think the extra verbosity of `case` helps the readability, especially when we're dealing with other `c`-things like courthouses etc.

I've noticed that IntelliJ complains with the alias `case` (is it a reserved JPQL keyword maybe?) But maybe we could use `courtCase` instead.

What do you think?

File: src/main/java/uk/gov/hmcts/darts/task/runner/impl/ApplyRetentionAutomatedTask.java (hunk 1 above)
Comment (Ben-Edwards-cgi) [reply src/main/java/uk/gov/hmcts/darts/task/runner/impl/ApplyRetentionAutomatedTask.java]: Updated :)

**Review by chrisbellingham-hmcts — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/darts/task/runner/impl/AbstractLockableAutomatedTask.java [hunk 3]
```diff
@@ -37,6 +38,10 @@
 @Slf4j
 public abstract class AbstractLockableAutomatedTask implements AutomatedTask, AutoloadingAutomatedTask {
 
+    private String taskPropertyName;
+    private Duration lockAtMostFor;
+    private Duration lockAtLeastFor;
```
[the remark this answers is not in the document]
Comment (Ben-Edwards-cgi) [reply src/main/java/uk/gov/hmcts/darts/task/runner/impl/AbstractLockableAutomatedTask.java:43]: Good spot :) These were left from some proto-typing I have removed

**Review by davet1985 — COMMENTED** [review]
Question on the task users

File: src/main/resources/application.yaml [hunk 4]
```diff
@@ -351,17 +351,159 @@ darts:
...
+          at-most-for: PT1H30M
+      unstructured-to-arm:
+        system-user-email: ${SYSTEM_USER_EMAIL:}
+        lock:
+          at-least-for: PT1M
+          at-most-for: PT1H30M
+
+      common-config:
+        thread-pool-size: 5
+        system-user-email: ${SYSTEM_USER_EMAIL:}
```
[the remark this answers is not in the document]
Comment (Ben-Edwards-cgi) [reply src/main/resources/application.yaml:503]: These were added in as they were apart of the common config, which all jobs require.
We can likely change these configuration items to match the email address used for that Job so the email is easily configurable

[the remark this answers is not in the document]
Comment (davet1985) [reply src/main/resources/application.yaml:503]: I don't think we need anything user related in the config? The users for each task are hardcoded in the code and don't need to be changed.

[the remark this answers is not in the document]
Comment (Ben-Edwards-cgi) [reply src/main/resources/application.yaml:503]: I don't think that ticket has been played yet as every job still requires the yaml system user value

[the remark this answers is not in the document]
Comment (Ben-Edwards-cgi) [reply src/main/resources/application.yaml:503]: image

See above in the abstract automated task code

**Review by hemantasharma1129 — APPROVED** [review]

File: src/main/resources/application.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (davet1985) [reply src/main/resources/application.yaml:503]: Yeah no problem, this can be addressed once that change is in.

File: src/main/java/uk/gov/hmcts/darts/arm/component/impl/AutomatedTaskProcessorFactoryImpl.java [hunk 5]
```diff
@@ -120,13 +120,4 @@ public GenerateCaseDocumentForRetentionDateProcessor createGenerateCaseDocumentF
             throw new DartsException(String.format("Batch size not supported for generating case document for retention date: '%s'", batchSize));
         }
     }
-
-    @Override
-    public CloseOldCasesProcessor createCloseOldCasesProcessor(int batchSize) {
```
Comment (jackmaloney) [inline src/main/java/uk/gov/hmcts/darts/arm/component/impl/AutomatedTaskProcessorFactoryImpl.java:125]: If you're removing this one, should you not remove the other ones in this class for consistency? Assuming there was a genuine reason for removing this.

File: src/main/java/uk/gov/hmcts/darts/casedocument/service/impl/GenerateCaseDocumentForRetentionDateBatchProcessorImpl.java [hunk 6]
```diff
@@ -33,11 +32,11 @@ public void processGenerateCaseDocumentForRetentionDate(int batchSize) {
         OffsetDateTime currentTimestamp = currentTimeHelper.currentOffsetDateTime();
         OffsetDateTime caseRetailUntilTimestamp = currentTimestamp.plusDays(caseDocumentExpiryDays);
         OffsetDateTime caseDocumentCreatedAfterTimestamp = currentTimestamp.minusDays(caseDocumentExpiryDays);
-        List<Integer> casesIds = caseRepository.findCasesNeedingCaseDocumentForRetentionDateGeneration(caseRetailUntilTimestamp,
-                                                                                                       caseDocumentCreatedAfterTimestamp,
-                                                                                                       Pageable.ofSize(batchSize));
+        var casesIds = caseRepository.findCasesNeedingCaseDocumentForRetentionDateGeneration(caseRetailUntilTimestamp,
```
Comment (jackmaloney) [inline src/main/java/uk/gov/hmcts/darts/casedocument/service/impl/GenerateCaseDocumentForRetentionDateBatchProcessorImpl.java:35]: Why was this changed to var? Not sure it meets the criteria for using var outlined in best practices here: https://tools.hmcts.net/confluence/pages/viewpage.action?pageId=1694706043#BestPractices(Backend)-Usevarthoughtfully

File: src/main/java/uk/gov/hmcts/darts/common/repository/CaseRepository.java (hunk 2 above)
Comment (jackmaloney) [reply src/main/java/uk/gov/hmcts/darts/common/repository/CaseRepository.java:29]: This was marked as resolved but seems like it's still there as `c`

File: src/main/resources/application.yaml [hunk 7]
```diff
@@ -355,14 +355,155 @@ darts:
       common-config:
         thread-pool-size: 5
         system-user-email: ${SYSTEM_USER_EMAIL:}
-      arm-rpo-duration: ${ARM_RPO_DURATION:1h}
-      inbound-annotation-transcription-deleter:
-        default-batch-size: 100000
+      apply-retention:
+        system-user-email: ${darts.automated.task.common-config.system-user-email:}
```
Comment (jackmaloney) [inline src/main/resources/application.yaml:359]: I take it these will all be updated as part of [DMP-4219](https://tools.hmcts.net/jira/browse/DMP-4219)?

File: src/main/resources/application.yaml [hunk 8]
```diff
@@ -355,14 +355,155 @@ darts:
...
+        minimum-duration-in-unstructured: 210d #10 weeks
+        lock:
+          at-least-for: PT1M
+          at-most-for: PT1H30M
+      unstructured-to-arm:
+        system-user-email: ${darts.automated.task.common-config.system-user-email:}
+        lock:
+          at-least-for: PT1M
+          at-most-for: PT1H30M
+      arm-rpo-duration: ${ARM_RPO_DURATION:1h}
```
Comment (jackmaloney) [inline src/main/resources/application.yaml:504]: Not sure the config around this part is right. It doesn't have the `system-user-email` and `lock` that other tasks use. Then the same for `process-e2e-arm-rpo`. Then you have `thread-pool-size:` and  `system-user-email` which don't look like they are in the right place, please can you review?

File: src/main/resources/application.yaml [hunk 9]
```diff
@@ -355,14 +355,155 @@ darts:
...
+      process-arm-response-files:
+        system-user-email: ${darts.automated.task.common-config.system-user-email:}
+        lock:
+          at-least-for: PT1M
+          at-most-for: PT1H30M
+      process-arm-rpo-pending:
+        system-user-email: ${darts.automated.task.common-config.system-user-email:}
+        lock:
+          at-least-for: PT1M
+          at-most-for: PT5H
```
Comment (jackmaloney) [inline src/main/resources/application.yaml:472]: The LLD has this as 60 mins

File: src/main/resources/application.yaml [hunk 10]
```diff
@@ -355,14 +355,155 @@ darts:
...
+      process-daily-list:
+        system-user-email: ${darts.automated.task.common-config.system-user-email:}
+        lock:
+          at-least-for: PT1M
+          at-most-for: PT5H
```
Comment (jackmaloney) [inline src/main/resources/application.yaml:477]: The LLD has this as 90 mins

File: src/main/resources/application.yaml [hunk 11]
```diff
@@ -373,12 +514,6 @@ darts:
       dets-to-arm-batch-processor:
```
Comment (jackmaloney) [inline src/main/resources/application.yaml:514]: Same comment as above, the config doesn't look right here.

File: src/main/resources/application.yaml (hunk 7 above)
Comment (Ben-Edwards-cgi) [reply src/main/resources/application.yaml:359]: Yes :) We can probably amend that ticket to just indicate that this config needs updating, this way anytime they grab the users identity the correct user is grabbed (Same for last modified by)

File: src/main/resources/application.yaml [hunk 12]
```diff
@@ -373,12 +514,6 @@ darts:
...
         max-arm-manifest-items: 100
         manifest-file-prefix: DETS
```
Comment (jackmaloney) [inline src/main/resources/application.yaml:516]: there are 3 tasks recently added to the LLD called `ArmRpoPoll`, `ArmRpoReplay` and `ArmMissingResponseReplay` that don't look like were included in this change. Could we add them?

File: src/main/resources/db/migration/common/V1_411__automated_task_batch.sql [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+update darts.automated_task set batch_size = 1000 where task_name = 'CaseExpiryDeletion';
+update darts.automated_task set batch_size = 5000, cron_expression = '0 0 0/1 ? * *' where task_name = 'AssociatedObjectDataExpiryDeletion';
+update darts.automated_task set batch_size = 200, cron_expression = '0 9 0,1,2,3,4,5,6,21,22,23 ? * *' where task_name = 'ProcessDETSToArmResponse';
+update darts.automated_task set batch_size = 5000, cron_expression = '0 0 21 * * ?' where task_name = 'ManualDeletion';
+update darts.automated_task set batch_size = 2000, cron_expression = '0 9 0,1,2,3,4,5,6,21,22,23 ? * *' where task_name = 'DetsToArm';
+update darts.automated_task set batch_size = 10000, cron_expression = '0 0 0,1,2,3,4,5,6,7,8,20,21,22,23 ? * *' where task_name = 'AudioLinking';
+update darts.automated_task set batch_size = 0, cron_expression = '0 18 23 * * ?' where task_name = 'ProcessE2EArmRpoPending';
+update darts.automated_task set batch_size = 10000, cron_expression = '0 0 0 31 2 *' where task_name = 'ArmRpoReplay';
+update darts.automated_task set batch_size = 10000, cron_expression = '0 */15 * ? * *' where task_name = 'ArmRpoPolling';
```
Comment (jackmaloney) [inline src/main/resources/db/migration/common/V1_411__automated_task_batch.sql:30]: Task ArmMissingResponseReplay missing from this script

File: src/main/resources/application.yaml [hunk 14]
```diff
@@ -355,14 +355,155 @@ darts:
...
+        system-user-email: ${darts.automated.task.common-config.system-user-email:}
+        lock:
+          at-least-for: PT1M
+          at-most-for: PT1H30M
+      audio-linking:
+        system-user-email: ${darts.automated.task.common-config.system-user-email:}
+        audio-buffer: 0s
+        lock:
+          at-least-for: PT1M
+          at-most-for: PT5H
```
[the remark this answers is not in the document]
Comment (Ben-Edwards-cgi) [reply src/main/resources/application.yaml:383]: Agreed updated :)

File: src/main/resources/db/migration/common/V1_411__automated_task_batch.sql (hunk 13 above)
Comment (jackmaloney) [inline src/main/resources/db/migration/common/V1_411__automated_task_batch.sql:30]: Trivial one, but the LLD has this listed as `ArmRpoPoll` - should we update the LLD to match what was implemented?

File: src/main/resources/application.yaml (hunk 9 above)
Comment (Ben-Edwards-cgi) [reply src/main/resources/application.yaml:472]: I dont see a job in the LLD for ProcessARMRPOPending but have updated this to be 90m

File: src/main/resources/application.yaml (hunk 10 above)
Comment (Ben-Edwards-cgi) [reply src/main/resources/application.yaml:477]: Agreed updated :)

File: src/main/resources/db/migration/common/V1_411__automated_task_batch.sql [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+update darts.automated_task set batch_size = 1000, cron_expression = '0 57 22 * * ?' where task_name = 'UnstructuredTranscriptionAnnotationDeleter';
+update darts.automated_task set batch_size = 5000, cron_expression = '0 50 0,1,2,3,4,5,6,7,8,20,21,22,23 ? * *' where task_name = 'GenerateCaseDocumentForRetentionDate';
+update darts.automated_task set batch_size = 1000 where task_name = 'CaseExpiryDeletion';
+update darts.automated_task set batch_size = 5000, cron_expression = '0 0 0/1 ? * *' where task_name = 'AssociatedObjectDataExpiryDeletion';
+update darts.automated_task set batch_size = 200, cron_expression = '0 9 0,1,2,3,4,5,6,21,22,23 ? * *' where task_name = 'ProcessDETSToArmResponse';
+update darts.automated_task set batch_size = 5000, cron_expression = '0 0 21 * * ?' where task_name = 'ManualDeletion';
+update darts.automated_task set batch_size = 2000, cron_expression = '0 9 0,1,2,3,4,5,6,21,22,23 ? * *' where task_name = 'DetsToArm';
+update darts.automated_task set batch_size = 10000, cron_expression = '0 0 0,1,2,3,4,5,6,7,8,20,21,22,23 ? * *' where task_name = 'AudioLinking';
```
Comment (jackmaloney) [inline src/main/resources/db/migration/common/V1_411__automated_task_batch.sql:27]: Can these cron expressions not be simplied? I think this has the same result as what you have here and is easier to read: 
`0 50 0-8,20-23 ? * *`

File: src/main/resources/application.yaml (hunk 9 above)
Comment (jackmaloney) [reply src/main/resources/application.yaml:472]: It's called `ProcessE2EArmRpoPending` in the LLD - I assume it's the same task @hemantasharma1129 ? We should probably update the LLD to align as i've noticed a few inconsistencies.

File: src/main/resources/application.yaml (hunk 8 above)
Comment (Ben-Edwards-cgi) [reply src/main/resources/application.yaml:504]: These are all existing unchanged configuration items.

However, I have moved some of these config items around to standardise there location

File: src/main/resources/application.yaml (hunk 12 above)
Comment (Ben-Edwards-cgi) [reply src/main/resources/application.yaml:516]: They were not included in this config as we do not yet have automated tasks made for them in code.
I would expect the config to be added once the automated task is added in code

File: src/main/resources/db/migration/common/V1_411__automated_task_batch.sql (hunk 13 above)
Comment (Ben-Edwards-cgi) [reply src/main/resources/db/migration/common/V1_411__automated_task_batch.sql:30]: This task does not exist in the code base yet so I would expect this to be included once the task is added

Comment (Ben-Edwards-cgi) [reply src/main/resources/db/migration/common/V1_411__automated_task_batch.sql:30]: Agreed @hemantasharma1129 please can you update the LLD

**Comment by jackmaloney** [comment]
How did this get merged with so many outstanding comments?

**Comment by Ben-Edwards-cgi** [comment]
As auto merge kicked in I have created https://github.com/hmcts/darts-api/pull/2267 to action these comments

File: src/main/resources/application.yaml (hunk 9 above)
Comment (hemantasharma1129) [reply src/main/resources/application.yaml:472]: I have added the missing job 'ProcessARMRPOPending' in the LLD. The implementation was done under DMP-3529.
